File: android/app/src/main/kotlin/org/getlantern/lantern/event/EventManager.kt
package org.getlantern.lantern.event

import android.os.Handler
import android.os.Looper
import io.flutter.embedding.engine.FlutterEngine
import io.flutter.plugin.common.EventChannel
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.flow.*
import kotlinx.coroutines.launch
import org.getlantern.lantern.model.Stats
import org.getlantern.mobilesdk.model.Event
import java.util.EnumMap
import java.util.concurrent.ConcurrentSkipListSet
import java.util.concurrent.atomic.AtomicReference

abstract class EventManager(
    private val name: String,
    flutterEngine: FlutterEngine? = null,
) : EventChannel.StreamHandler {
    private val activeSubscribers: EnumMap<Event, ConcurrentSkipListSet<Int>> =
        EnumMap(Event::class.java)

    private val activeSink = AtomicReference<EventChannel.EventSink?>()

    private val handler = Handler(Looper.getMainLooper()) // a handler for the main thread

    init {
        flutterEngine?.let {
            EventChannel(
                flutterEngine.dartExecutor,
                name,
            ).setStreamHandler(this)
        }
    }

    fun onNewEvent(
        event: Event,
        params: MutableMap<String, Any?> = mutableMapOf(),
    ) {
        handler.post {
            synchronized(this@EventManager) {
                params["eventName"] = event.name

                // notify subscribers who listens for only 1 specific event
                var subscribers = activeSubscribers[event]
                subscribers?.forEach { subscriberID ->
                    params["subscriberID"] = subscriberID
                    activeSink.get()?.success(params)
                }

                // notify subscribers who listens for all events
                subscribers = activeSubscribers[Event.All]
                subscribers?.forEach { subscriberID ->
                    params["subscriberID"] = subscriberID
                    activeSink.get()?.success(params)
                }
            }
        }
    }

    suspend fun postStatsEvent(stats: Stats) {
        CoroutineScope(Dispatchers.IO).launch {
            EventBus.publish(AppEvent.StatsEvent(stats))
        }
    }

    fun subscribeAppEvent(onEvent: (AppEvent) -> Unit) {
        CoroutineScope(Dispatchers.Main + SupervisorJob()).launch {
            EventBus.subscribe<AppEvent> { appEvent ->
                onEvent(appEvent)
            }
        }
    }

    @Synchronized
    override fun onListen(
        arguments: Any?,
        events: EventChannel.EventSink?,
    ) {
        activeSink.set(events)
        val args = arguments as Map<String, Any>
        val subscriberID = args["subscriberID"] as Int
        val eventName = args["eventName"] as String
        val event = Event.valueOf(eventName)

        val subscribers = activeSubscribers[event] ?: ConcurrentSkipListSet<Int>()
        subscribers.add(subscriberID)
        activeSubscribers[event] = subscribers
        onListen(event)
    }

    abstract fun onListen(event: Event)

    override fun onCancel(arguments: Any?) {
        if (arguments == null) {
            return
        }
        val args = arguments as Map<String, Any>
        val subscriberID = args["subscriberID"] as Int
        for (entry in activeSubscribers.entries) {
            entry.value.remove(subscriberID)
        }
    }
}

File: ios/Runner/Lantern/EventManager.swift
//
//  EventManager.swift
//  Runner
//
//  Created by jigar fumakiya on 14/07/23.
//

import Foundation
import Flutter


enum Event: String, CaseIterable {
    case all = "All"
    case surveyAvailable = "SurveyAvailable"
    case noNetworkAvailable = "NoNetworkAvailable"
    case networkAvailable = "NetworkAvailable"
    case noProxyAvailable = "NoProxyAvailable"
    case proxyAvailable = "ProxyAvailable"
}



class EventManager: NSObject, FlutterStreamHandler {
    private let name: String
    private var activeSubscribers: [Event : Set<Int>] = [:]
    private var activeSink: FlutterEventSink?
    private let mainThreadHandler = DispatchQueue.main
    var onListenClosure: (Event) -> Void

    init(name: String, binaryMessenger: FlutterBinaryMessenger ,onListenClosure: @escaping ((Event) -> Void)) {
        self.onListenClosure = onListenClosure
        self.name = name
        super.init()
        let eventChannel = FlutterEventChannel(name: name, binaryMessenger: binaryMessenger)
        eventChannel.setStreamHandler(self)
        
    }
    
    func onNewEvent(event: Event, params: [String : Any] = [:]) {
        var params = params
        mainThreadHandler.async {
            self.activeSubscribers[event]?.forEach { subscriberID in
                params["subscriberID"] = subscriberID
                self.activeSink?(params)
            }
            
            self.activeSubscribers[.all]?.forEach { subscriberID in
                params["subscriberID"] = subscriberID
                self.activeSink?(params)
            }
        }
    }
  

    // MARK: - FlutterStreamHandler
    func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        activeSink = events
        if let args = arguments as? [String: Any],
            let subscriberID = args["subscriberID"] as? Int,
            let eventName = args["eventName"] as? String,
            let event = Event(rawValue: eventName) {
            
            let subscribers = activeSubscribers[event] ?? []
            activeSubscribers[event] = subscribers.union([subscriberID])
            onListenClosure(event)
        }
        return nil
    }
    
    func onCancel(withArguments arguments: Any?) -> FlutterError? {
        if let args = arguments as? [String: Any], let subscriberID = args["subscriberID"] as? Int {
            activeSubscribers.forEach { event, subscriberSet in
                activeSubscribers[event]?.remove(subscriberID)
            }
        }
        return nil
    }
    

}
